### ISSUE TITLE
cl-dbi:commit should make a non-local exit

### ISSUE THREAD

**Issue by svetlyak40wt** [fukamachi/cl-dbi#49]
This behavior is implicit, unexpected and leads to many WTFs.

This is a short illustration:

```lisp
CL-USER> (dbi:with-transaction mito.connection:*connection*
           (format t "BEFORE COMMIT~%")
           (cl-dbi:commit mito:*connection*)
           (format t "AFTER COMMIT~%"))
BEFORE COMMIT
T
```

Code after `cl-dbi:commit` have no change to be executed! This is because cl-dbi:commit signals `dbi.driver::transaction-done-condition` and it is catched in `dbi:with-transaction` by `handler-case`:

```lisp
(let ((#:conn-var1322 mito.connection:*connection*))
  (if (find #:conn-var1322 dbi.driver:*in-transaction* :test #'eq)
      (dbi:with-savepoint #:conn-var1322
        (format t "BEFORE COMMIT~%")
        (dbi.driver:commit mito.connection:*connection*)
        (format t "AFTER COMMIT~%"))
      (let* (#:transaction-done1323
             #:transaction-ok1324
             (#:conn-var1325 #:conn-var1322)
             (dbi.driver:*in-transaction*
              (cons #:conn-var1325 dbi.driver:*in-transaction*)))
        (dbi.driver:begin-transaction #:conn-var1325)
        (unwind-protect
            (handler-case
             (multiple-value-prog1
                 (progn
                  (format t "BEFORE COMMIT~%")
                  (dbi.driver:commit mito.connection:*connection*)
                  (format t "AFTER COMMIT~%"))
               (setf #:transaction-ok1324 t))
             (dbi.driver::transaction-done-condition nil
              (setf #:transaction-done1323 t)))
          (unless #:transaction-done1323
            (handler-case
             (if #:transaction-ok1324
                 (dbi.driver:commit #:conn-var1325)
                 (dbi.driver:rollback #:conn-var1325))
             (dbi.driver::transaction-done-condition nil)))))))
```

Probably, `handler-bind` could be used here to keep track if transaction was commited.

**Comment by svetlyak40wt** [comment]
I've experimented with this code a little bit. Probably the better way will be to refactor handling transactions and savepoints and to keep a transaction's state in `*in-transaction*` list like:

```lisp
(defclass transaction-state ()
    ((conn :documentation "Connection where transaction was started")
     (state :documentation "Either transaction in progress or not"
                :type (member :in-progress :commited :rolled-back)))
```

**Comment by svetlyak40wt** [comment]
@fukamachi if you lack of time, I could make a pull request.

**Comment by fukamachi** [comment]
This change is intended not to run more SQLs after commit/rollback because
I think that those operations are manual explicit terminations of
with-transaction.

2019年10月4日(金) 15:59 Alexander Artemenko <notifications@github.com>:

**Comment by fukamachi** [comment]
It could be done with `handler-bind`, as you mentioned, and raise an error for following `execute-with-connection` or `do-sql`.

**Comment by svetlyak40wt** [comment]
My case was like that:

I have a web request handler which has around method with `with-transaction` call. Here is simplified pseudocode describing my problem:

```lisp
(define-condition immediate-response ()
  ((html :initarg :html)
   (code :initarg :code))
  (:documentation "Raise this to return response from any stage of the request."))

(defgeneric handle-request (app)
  (:documentation "This function should return a string with HTML response."))

(defmethod handle-request :around ((app t))
  (dbi:with-transaction (dbi:*connection*)
    (handler-case (call-next-method)
      (immediate-response (condition)
        (when (< (get-code condition)
                 500)
          ;; when code was OK, I want to save the state to the database
          (dbi:commit))
        ;; And in any case to return HTML, but right now
        ;; cl-dbi:commit prevent this line's execution because
        ;; it throws transaction-done-condition
        (get-html condition)))))

(defclass my-app ()
  ())


(defun find-the-object ()
  (let ((results (execute "SELECT * FROM objects")))
    (unless results
      (signal 'immediate-response
              :code 404
              :html "Object not found"))))


(defmethod handle-request ((app my-app))
  (let ((object (find-the-object)))
    (do-the-job-with object)
    (render-html object)))
```

**Comment by svetlyak40wt** [comment]
I'll try to make a pull if you don't mind.

**Comment by fukamachi** [comment]
Now I see a problem you have and I admit that it’s an unpleasant behavior.

> I'll try to make a pull if you don't mind.
Of course not. I appreciate your contribution.

**Comment by svetlyak40wt** [comment]
@fukamachi please, review this pull-request:

https://github.com/fukamachi/cl-dbi/pull/50

**Comment by fukamachi** [comment]
Probably fixed by #50. Thanks a lot.
